### PR TITLE
DEV-3164 | Standardize page view tests and retire a bunch of technical debt

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

This PR is into a feature branch (DEV-2718), not main. This PR should be merged, and following that #939.

#### What's this PR do?

This PR removes a bunch of cruft around object-level permissions that stemmed from not appreciating nuances of DRF at original time of coding.  It also refactors our tests (primarily in the pages app) to remove unittest, only use Pytest, and take advantage of parametrized pytest fixtures.

Some of the notable specific changes:

- Refactors `HasRoleAssignment.has_permission` to only allow through users who have an expected role type. This in turn allowed a couple of tests that had curiously been asserting for presence of 500 to now start looking for 403, which  makes more sense for case.
- Retires `_BaseAssumedViewAndModelMixin` and `HasDeletePrivilegesViaRole`. These were creating a bunch of indirection and were there because I originally didn't understand there was a much more straightforward way to determine delete privileges by retrieving a queryset filtered for the user's roleassignment (via a custom model manager method) ahead of deleting (and remaining entirely within DRF views and serializers, without needing custom classes).
- Refactors the `ValidateFkReferenceOwnership` to take a `model` attribute and to use the model's `filtered_by_role_assignment` method.
- Retires the old, idiosyncratic `organizations.models.Organization.filter_queryset_by_role_assignment` in favor of a new `organizations.models.OrganizationManager.filtered_by_role_assignment` method (which is the approach we're trying to standardize across the app).
- Adds new traits to our OrganizationFactory for testing. The`free_plan` and `plus_plan` boolean attributes can be passed at factory initialization time to create an org on either plan.
- Adds a new `onboarded` trait to the `PaymentProviderFactory`. By setting this to true at init time, a provider's state will mimic that of a real one whose user has gone through Stripe onboarding process.
- Some refactoring of `test_org_views.py` to use common fixtures exposed in conftest.py and in light of removal of aforementioned crufty mixins etc.
- Adds `PageManager` and `StyleManager` model managers to `DonationPage` and `Style` models respectively.  Both managers have a `filtered_by_role_assignment` method that filters results based on characteristics of passed role assignment.
- Removes `AbstractPage` model entirely, as retiring `Template` model means there's only one inheriting class `DonationPage`, which now simply itself contains relevant attributes and functions from `AbstractPage`.
- Makes validation of sidebar and page elements stricter (in apps/pages/serializers.py) 
- Adds a `published` trait to `DonationPageFactory` which will put the page in a published state.
- Refactors the `StyleFactory` to have a representative value for the `styles` property (which our serialization layer requires to have certain properties that were not represented in the factory until now).
- Adds a `DonationPageFullDetailSerializer.validate_revenue_program` to account for RP being optional for updates but required for creation.
- Fixes the name-incrementing behavior for page name. When a page name is taken, if rp name is available as a page name, it should be used first without appending a `1`. Only append when orig is taken.
- Revises`.coveragerc` so that settings files are actually ignored (there was previously a line that intended but failed to do this). Also removes the broken config to ignore management command files since that line was broken and in fact we are (and should) test management commands.
- Adds `apps.common.views.FilterForSuperUserOrRoleAssignmentUserMixin` which drys up repeated logic shared by views where queryset should be prepped for a super user or user with role assignment.
- Retires `apps.users.models.RoleAssignmentResourceModelMixin` as it is no longer needed.
- Makes some non-db-tracked changes to old migrations that referenced `apps.users.models.RoleAssignmentResourceModelMixin`. Not sure why this mixin ended up hard coded in migrations -- perhaps because it should have subclassed django model?. At any rate, this change to migrations seems harmless, and lowest effort / cleanest way to deal with this hard coded reference disappearing and being unavailable in migrations.
- Refactors contributions models to use custom queryset + manager for `filter_queryset_for_contributor` and `filtered_by_role_assignment` methods.
- Refactors many though not all of the tests for apps.contributions.views to use emergent standard approach (pytest + parametrized fixtures instead of unitest and `set_up_domain_model`).

#### Why are we doing this? How does it help us?

- Remove a bunch of technical debt
- Better testing
- Better dev experience
- Eliminate big source of "bad pattern examples" that don't want other devs extending
- Gets us much closer to pytest-only world, in which we can turn test parallelization to improve test run speed

#### How should this be manually tested? Please include detailed step-by-step instructions.

Normal QA.  All original features should work.

The test coverage is extensive, and has only expanded with this PR, and there are a handful of small bugs or inconsistencies in API permissioning behavior that it solved for (things not previously tracked in tickets, but edge cases that arose as unsolved when rewriting tests as pytest tests using parametrized fixtures).   While most of this PR is test changes, there is a layer of permissions-related cruft that was removed. While the tests are good evidence this has not caused any regressions, in QA, we should ensure that page and style create/edit/delete flows work in the org dashboard.  We should also ensure that public facing contribution pages work as expected. We don't need to test a bunch of edge cases, but just normal flow through page creation, and contribution checkout flow.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No. This PR removes a *lot* of tech debt.

#### Has this been documented? If so, where?

No. There's already a ticket to document emergent backend testing strategy.

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-3164](https://news-revenue-hub.atlassian.net/browse/DEV-3164)
- [DEV-2718](https://news-revenue-hub.atlassian.net/browse/DEV-2718)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

[DEV-3150 | Document philosophy on testing API layer of Django App](https://news-revenue-hub.atlassian.net/browse/DEV-3150)

[DEV-3164]: https://news-revenue-hub.atlassian.net/browse/DEV-3164?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/pages/tests/test_viewsets.py [hunk 1]
```diff
@@ -1,1292 +1,1872 @@
...
+
+@pytest.fixture
+def style():
+    return StyleFactory()
+
+
+@pytest.fixture
+def patch_page_valid_data():
+    return {
+        "name": fake.company(),  # random
```
Comment (benjaminEwhite) [inline apps/pages/tests/test_viewsets.py:128]:
```suggestion
        "name": fake.company(),
```

File: apps/contributions/tests/test_views.py [hunk 2]
```diff
@@ -287,9 +287,7 @@ def test_contributions_are_read_only_for_expected_users(self):
 
     def test_unexpected_role_type(self):
         novel = create_test_user(role_assignment_data={"role_type": "never-before-seen"})
-        self.assert_user_cannot_get(
-            reverse("contribution-list"), novel, expected_status_code=status.HTTP_500_INTERNAL_SERVER_ERROR
-        )
+        self.assert_user_cannot_get(reverse("contribution-list"), novel, expected_status_code=status.HTTP_403_FORBIDDEN)
```
Comment (benjaminEwhite) [inline apps/contributions/tests/test_views.py:290]: This change is because we now get a more sensical status code after solving for a bug that caused an uncaught 500 when unexpected role type.

File: apps/pages/tests/factories.py [hunk 3]
```diff
@@ -35,7 +32,11 @@ class Params:
         org = None
 
     name = factory.Sequence(lambda n: "Test Style %d" % n)
-    styles = {"colors": {"primary": "testing-pink"}}
+    styles = {
```
Comment (benjaminEwhite) [inline apps/pages/tests/factories.py:35]: Previously, the value set for `styles` here was not one that validation would allow. Below constitutes a minimally valid `styles` value.

File: apps/pages/tests/test_models.py [hunk 4]
```diff
@@ -7,22 +7,17 @@
...
-    DefaultPageLogo,
-    DonationPage,
-    Style,
-    Template,
-    _get_screenshot_upload_path,
-)
+from apps.pages.models import DefaultPageLogo, DonationPage, Style, _get_screenshot_upload_path
 from apps.pages.tests.factories import DonationPageFactory, FontFactory, StyleFactory
-from apps.users.choices import Roles
+from apps.users.models import Roles
```
Comment (benjaminEwhite) [inline apps/pages/tests/test_models.py:20]:
```suggestion
from apps.users.choices import Roles
```

**Review by njharman — CHANGES_REQUESTED** [review]
Overall great direction and code clean up.  A scattering of minor things and suggestions. And...

pre-commit failed, safety, complaining about lib "future" vulnerability.  Also "Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them."

Those don't seem to fall under ticket disclaimer about failing tests due to cypress.

File: apps/common/validators.py [hunk 5]
```diff
@@ -22,6 +22,7 @@ def _has_ensured_user_ownership_by_default(user, role_assignment):
     )
 
 
+# todo: make this more "Django-esque"
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/common/validators.py]: Removed. 5eeac967878e8

File: apps/common/validators.py [hunk 6]
```diff
@@ -47,6 +49,11 @@ def __init__(self, fk_attribute, has_default_access_fn=_has_ensured_user_ownersh
                 has_default_access_fn.__code__.co_varnames,
             )
             raise serializers.ValidationError(f"{self.__class__} initialized with function with unexpected signature")
+        if not getattr(self.model.objects, "filtered_by_role_assignment", None):
+            # logger.warning(f"")
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/common/validators.py]: 743c28d6

File: apps/pages/models.py [hunk 7]
```diff
@@ -186,31 +113,21 @@ def save(self, *args, **kwargs):
...
+    def filtered_by_role_assignment(self, role_assignment: RoleAssignment) -> models.QuerySet:
+        match role_assignment.role_type:
+            case Roles.HUB_ADMIN:
+                return self.all()
+            case Roles.ORG_ADMIN:
+                return self.filter(revenue_program__organization=role_assignment.organization)
+            case Roles.RP_ADMIN:
+                return self.filter(revenue_program__in=role_assignment.revenue_programs.all())
+            case _:
+                return self.none()
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/models.py]: Dried up into a common manager used in both models.

1d6024a6a0a

File: apps/pages/serializers.py [hunk 8]
```diff
@@ -75,19 +76,27 @@ def to_internal_value(self, data):
...
         """
-        name = data.pop("name", None)
-        revenue_program = data.pop("revenue_program", None)
-        data = {
-            "name": name,
-            "revenue_program": revenue_program,
-            "styles": data,
-        }
-        return super().to_internal_value(data)
+        name = data.pop("name", self.instance.name if self.instance else None)
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/serializers.py:81]: 815c3b28c1

File: apps/pages/serializers.py [hunk 9]
```diff
@@ -242,19 +242,22 @@ def validate_page_limit(self, data):
     def validate_page_element_permissions(self, data):
         """Ensure that requested page elements are permitted by the organization's plan"""
         rp = self.instance.revenue_program if self.instance else data["revenue_program"]
-        if prohibited := {x["type"] for x in data.get("elements", [])}.difference(
-            set(rp.organization.plan.page_elements)
-        ):
+        prohibited = []
+        if any([x.get("type", None) is None for x in data.get("elements", [])]):
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/serializers.py]: 82b861fd36

File: apps/pages/serializers.py [hunk 10]
```diff
@@ -242,19 +242,22 @@ def validate_page_limit(self, data):
...
             )
 
     def validate_sidebar_element_permissions(self, data):
         """Ensure that requested sidebar elements are permitted by the organization's plan"""
         rp = self.instance.revenue_program if self.instance else data["revenue_program"]
-        if prohibited := set([elem["type"] for elem in data.get("sidebar_elements", [])]).difference(
-            set(rp.organization.plan.sidebar_elements)
-        ):
+        if prohibited := set(
+            [elem["type"] for elem in data.get("sidebar_elements", []) if elem.get("type", None)]
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/serializers.py]: bb26bbf0f0

File: apps/organizations/views.py [hunk 11]
```diff
@@ -49,17 +46,18 @@ class OrganizationViewSet(
...
 
-    def get_queryset(self):
+    def get_queryset(self) -> models.QuerySet:
         if self.request.user.is_superuser:
             return Organization.objects.all()
         elif ra := self.request.user.get_role_assignment():
             return Organization.objects.filtered_by_role_assignment(ra)
-        else:
+        # given permission clases, this may be unreachable, but logically gaurantees this returns a queryset
+        else:  # pragma: no cov
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/views.py:55]: Weird. Comes up as uncovered even with no branch. Not sure what's going on here, but possibly a bug with Coverage.

File: apps/pages/serializers.py [hunk 12]
```diff
@@ -242,19 +242,22 @@ def validate_page_limit(self, data):
     def validate_page_element_permissions(self, data):
         """Ensure that requested page elements are permitted by the organization's plan"""
         rp = self.instance.revenue_program if self.instance else data["revenue_program"]
-        if prohibited := {x["type"] for x in data.get("elements", [])}.difference(
-            set(rp.organization.plan.page_elements)
-        ):
+        prohibited = []
+        if any([x.get("type", None) is None for x in data.get("elements", [])]):
+            raise serializers.ValidationError({"page_elements": "Something is wrong with the provided page elements"})
+        if prohibited := [
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/serializers.py]: db9345bad

File: apps/organizations/views.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/views.py]: 🤦 it needed to be `# pragma: no cover`

638131f12

File: apps/pages/views.py [hunk 13]
```diff
@@ -262,8 +234,14 @@ class StyleViewSet(RevisionMixin, viewsets.ModelViewSet, FilterQuerySetByUserMix
...
 
+    # prohibit put
+    http_method_names = ["get", "post", "patch", "delete"]
+
     def get_queryset(self):
-        return self.filter_queryset_for_user(self.request.user, self.model.objects.all())
+        if self.request.user.is_superuser:
+            return self.model.objects.all()
+        else:
+            return self.model.objects.filtered_by_role_assignment(self.request.user.roleassignment)
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/pages/views.py]: Good point. 827dcb748

**Comment by benjaminEwhite** [comment]
@njharman responded to all in-line comments.

Re: 

> pre-commit failed, safety, complaining about lib "future" vulnerability. Also "Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them."

I've merged in upstream changes and the `future` vulnerability is gone. 

I did `poetry update` which should take care of the other issue (and didn't seem otherwise problematic, but let me know if this is a bad idea): 6688a034ad

**Review by njharman — APPROVED** [review]
Looks good, ship it!

Thanks for ticket comment highlighting changed code (and why). It was very useful in helping me figure out what to look at for rereview.

**Review by x110dc — APPROVED** [review]
Great to see so much code removed. LGTM!
